tests/conftest.py

```diff
@@ -220,53 +232,122 @@ def network_config():
     yield ipv4_generator
 
 
-@pytest.fixture
-def microvm_image_fetcher():
-    """Return a borg object that knows about fetching microvm images.
-
-    If `ENV_TEST_IMAGES_S3_BUCKET` is set in the environment, target the bucket
-    specified therein, else use the default.
-    """
-    if ENV_TEST_IMAGES_S3_BUCKET in os.environ:
-        test_images_s3_bucket = os.environ.get(ENV_TEST_IMAGES_S3_BUCKET)
-    else:
-        test_images_s3_bucket = DEFAULT_TEST_IMAGES_S3_BUCKET
-
-    return MicrovmImageS3Fetcher(test_images_s3_bucket)
-
-
 @pytest.fixture(
-    params=microvm_image_fetcher().list_microvm_images(
+    params=MICROVM_S3_FETCHER.list_microvm_images(
         capability_filter=['*']
     )
 )
-def test_microvm_any(request, microvm, microvm_image_fetcher):
+def test_microvm_any(request, microvm):
     """Yield a microvm that can have any image in the spec bucket.
 
     A test case using this fixture will run for every microvm image.
 
     When using a pytest parameterized fixture, a test case is created for each
     parameter in the list. We generate the list dynamically based on the
     capability filter. This will result in
-    `len(microvm_image_fetcher.list_microvm_images(capability_filter=['*']))`
+    `len(MICROVM_S3_FETCHER.list_microvm_images(capability_filter=['*']))`
     test cases for each test that depends on this fixture, each receiving a
     microvm instance with a different microvm image.
     """
     # pylint: disable=redefined-outer-name
     # The fixture pattern causes a pylint false positive for that rule.
 
-    microvm_image_fetcher.get_microvm_image(request.param, microvm)
+    MICROVM_S3_FETCHER.init_vm_resources(request.param, microvm)
     yield microvm
 
 
+@pytest.fixture
+def test_multiple_microvms(
+        test_session_root_path,
+        context,
+        newpid_cloner_path
+):
+    """Yield a microvm or multiple microvms based on the context provided.
+
+    `context` is a dynamically parameterized fixture created inside the special
+    function `pytest_generate_tests` and it holds a tuple containing the name
+    of the guest image used to spawn a microvm and the number of microvms
+    to spawn.
+    """
+    # pylint: disable=redefined-outer-name
+    # The fixture pattern causes a pylint false positive for that rule.
+    microvms = []
+    (microvm_resources, how_many) = context
+
+    # When the context specifies multiple microvms, we use the first vm to
+    # populate the other ones by hardlinking its resources.
+    first_vm = init_microvm(test_session_root_path, newpid_cloner_path)
+    MICROVM_S3_FETCHER.init_vm_resources(
+        microvm_resources,
+        first_vm
+    )
+    microvms.append(first_vm)
+
+    for _ in range(how_many - 1):
```

Comment:
I noticed that the `context` `assert`s on `how_many` to make sure it's always > 0, can you add a comment here about that so that everyone can be positive when reading this code that `how_many` is always valid?

Reply:
Done.